File: fsharp/simple-cipher/SimpleCipher.fs
module SimpleCipher

type SimpleCipher(key: string) =
    let shiftEncode shiftChar c = 
        let zero = 'a' |> int
        let shiftKey = (shiftChar |> int) - zero
        char <| (int c - zero + shiftKey) % 26 + int zero 
        
    let shiftDecode shiftChar c = 
        let zero = 'a' |> int
        let shiftKey = (shiftChar |> int) - zero
        char <| (int c - zero - (shiftKey - 26) ) % 26 + int zero 
    let modulo = key |> String.length

    new() = 
        let rand = new System.Random()
        let randKey = [0..99] |> Seq.map (fun _ -> rand.Next(26) + ('a' |> int) |> char) |> Seq.toArray |> System.String
        new SimpleCipher(randKey)

    member val Key = key
        with get
       
    member this.Encode(plaintext: string) = 
        plaintext
        |> Seq.mapi (fun i c -> shiftEncode (this.Key.[i % modulo]) c)
        |> Seq.toArray
        |> System.String

    member this.Decode(ciphertext: string) = 
        ciphertext
        |> Seq.mapi (fun i c -> shiftDecode (this.Key.[i % modulo]) c)
        |> Seq.toArray
        |> System.String
    

File: csharp/simple-cipher/SimpleCipher.cs
using System;
using System.Linq;

public class SimpleCipher
{
    private string _key;
    public SimpleCipher()
    {
        Random rand = new Random();
        int zero = 'a';
        _key = new string(Enumerable.Range(0, 100).Select(_ => (char)(rand.Next(26) + zero)).ToArray());
    }
    public SimpleCipher(string key)
    {
        _key = key;
    }
    public string Key
    {
        get
        {
            return _key;
        }
    }
    public string Encode(string text)
    {
        int repetitions = text.Length / Key.Length + 1;
        string longerKey = String.Concat(Enumerable.Repeat(Key, repetitions));
        char[] chars = text.Zip(longerKey, (c, k) => char.IsLetter(c) ? ShiftEncode(c, (int)k - (int)'a') : c).ToArray();
        return new string(chars);
    }
    public string Decode(string text)
    {
        int repetitions = text.Length / Key.Length + 1;
        string longerKey = String.Concat(Enumerable.Repeat(Key, repetitions));
        char[] chars = text.Zip(longerKey, (c, k) => char.IsLetter(c) ? ShiftDecode(c, (int)k - (int)'a') : c).ToArray();
        return new string(chars);
    }
    public static char ShiftEncode(char c, int shift)
    {
        int zero = 'a';
        return (char)(((c - zero) + shift + 26) % 26 + zero);
    }
    public static char ShiftDecode(char c, int shift)
    {
        int zero = 'a';
        return (char)(((c - zero) - shift + 26) % 26 + zero);
    }
}